asynch/pool.py

```diff
@@ -144,23 +144,42 @@ async def _create_connection(self) -> None:
         await conn.connect()
         self._free_connections.append(conn)
 
-    async def _acquire_connection(self) -> Connection:
+    def _pop_connection(self) -> Connection:
         if not self._free_connections:
             raise AsynchPoolError(f"no free connection in {self}")
 
-        conn = self._free_connections.popleft()
+        return self._free_connections.popleft()
+
+    async def _acquire_connection(self) -> Connection:
+        conn = self._pop_connection()
+
+        try:
+            await conn._refresh()
+        except ConnectionError:
+            await self._create_connection()
+            conn = self._pop_connection()
```

Comment:
In the second call to `_pop_connection()`, are we guaranteed to get the recently created connection, or could we end up with another stale connection?

Wouldn't we have to pop connections in a loop until we get one that's fresh? Ie sth like:

```python
while True:
    conn = pop_connection()

    try:
        conn.refresh()
        return conn
    except:
        pass

```

Reply:
Ok, let's follow the train of steps:

157. connection refreshing failed -> ConnectionError
158. intercepting the raised exception
159. attempting to create a new connection as a replacement -> and if it fails, the exception is propagated, otherwise, a "successful success", but the connection "vitality" cannot be guaranteed, but it fills the ranks of the free connection candidates
160. popping a connection - the given one may not be that from the previous line, let's assume an already broken connection to be issued

Next, we shall bear in mind that all connections will be invalidated.

Moving on. When leaving the pool, the broken connection will be inoperative and let's say the standard procedure is to leave the compromised context -> moving to the `_release_connection` [snippet]( https://github.com/long2ice/asynch/pull/130/files#diff-595505f30f8dea14e55f38d8cd4975897b5e28f33f53205a94ff6f42d9003148R166-R178 ):
- attempting to refresh the acquired connection -> failure -> NOT putting back this connection into the registry of free connections
- removing from the registry of acquired connections -> it is right and must be done right because it gets released -> not acquired any more
- attempting to create a connection -> this won't do any miracle and the exception waves its hand (elt's imagine), BUT -1 acquired connection as it should be and no rotten connection to increment the free ones - the pool lost one but invalid connection

There are major questions:
1. Can we acquire a connection from the pool? I admit that the pool should do its best to find any valid connection before attempting to create new ones -> this is about "to be" and this is what you ( @nils-borrmann-tacto ) pointed out. The "as is" results in exception when recreation fails.
2. When the pool must be repopulated with `pool.minsize` connections? Now it is when acquiring a connection ([this cbranch](https://github.com/long2ice/asynch/pull/130/files#diff-595505f30f8dea14e55f38d8cd4975897b5e28f33f53205a94ff6f42d9003148R213) and it should be immediately after [this line](https://github.com/long2ice/asynch/pull/130/files#diff-595505f30f8dea14e55f38d8cd4975897b5e28f33f53205a94ff6f42d9003148R229) under the effect of the pool lock.

Thank you for your ideas, I saw other imperfections and weak parts.

Reply:
Implemented some ideas, please appreciate if you can.